packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -308,29 +321,41 @@ contract Bridge is EssentialContract, IBridge {
                     // time of the call.
                     //
                     // See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md
-                    if (_message.gasLimit > (gasleft() * 63) >> 6) revert B_NOT_ENOUGH_GASLEFT();
 
-                    gasLimit = _message.gasLimit;
+                    if (invocationGasLimit > (gasleft() * 63) >> 6) revert B_NOT_ENOUGH_GASLEFT();
                 }
 
-                if (_invokeMessageCall(_message, msgHash, gasLimit)) {
+                if (_invokeMessageCall(_message, msgHash, invocationGasLimit)) {
                     _updateMessageStatus(msgHash, Status.DONE);
                 } else {
                     _updateMessageStatus(msgHash, Status.RETRIABLE);
                 }
             }
 
-            // Determine the refund recipient
-            address refundTo =
-                _message.refundTo == address(0) ? _message.destOwner : _message.refundTo;
-
             // Refund the processing fee
-            if (msg.sender == refundTo) {
-                refundTo.sendEtherAndVerify(_message.fee + refundAmount, _SEND_ETHER_GAS_LIMIT);
+            if (msg.sender == _message.destOwner) {
+                _message.destOwner.sendEtherAndVerify(
+                    _message.fee + refundAmount, _SEND_ETHER_GAS_LIMIT
+                );
             } else {
+                // gas now represents the actual gas used
+                uint256 fee;
+                {
+                    gasAmount -= gasleft();
+                    gasAmount += isNewlyProven
+                        ? ONE_STEP_PROCESSING_GAS_OVERHEAD
+                        : TWO_STEP_PROCESSING_GAS_OVERHEAD;
+
+                    uint256 maxFee = _message.fee * gasAmount / _message.gasLimit;
```

Comment:
_messsage.fee (in previous versions at least) represented the  the "whole" fee the relayer pocket. In this context it is more `fee_per_gas`, no ?

Reply:
I think it would still be the whole fee, and `* (gasAmount / _message.gasLimit)` just takes the fraction of the max total fee.

Reply:
_message.fee: max fee amount in ETH
_message.gasLimit: the tx's max gas limit including the overhead
_message.gasLimit - TWO_STEP_PROCESSING_OVERHEAD: the function call's max gas limit
_message.fee/_message.gasLimit: the max fee per gas.